src/misc/ftwx.3
.\"(c) Copyright 1992 by Panagiotis Tsirigotis
.\"All rights reserved.  The file named COPYRIGHT specifies the terms
.\"and conditions for redistribution.
.\"
.\" $Id: ftwx.3,v 1.2 2003/06/17 05:10:52 seth Exp $
.TH FTWX 3X "25 January 1992"
.SH NAME
ftwx -- file tree walk
.SH SYNOPSIS
.LP
.nf
.ft B
#include "ftwx.h"
.LP
.ft B
.ft B
.ft B
int ftwx( path, func, depth, flags )
char *path ;
int (*func)() ;
int depth ;
int flags ;
.SH DESCRIPTION
.LP
.B ftwx()
is an extension to \fIftw(3)\fR that optionally follows symbolic
links (the default is not to follow them). It recursively descends
the directory tree whose root is \fIpath\fR. For each object it finds
(an object is anything that is in the file system name space,
like a directory, a socket, a regular file, a device etc.)
it invokes \fIfunc()\fR passing
it a pointer to a string containing the object path name (the first
component of the path will be \fIpath\fR), a pointer to a \fIstat()\fR
structure that contains information about the object, and a flag (of
type \fIint\fR). The user function is never invoked for the directory
entries "." and "..".
Possible flag values are:
.TP 10
.SB FTW_D
the object is a directory
.TP
.SB FTW_DNR
the object is a directory that is not readable
.TP
.SB FTW_F
the object is not a directory
.TP
.SB FTW_NS
the \fIstat()\fR of the object failed. The value of the pointer passed
to the user function is undefined. This flag value will be returned
for directory entries of a directory for which execute access is not
permitted.
.LP
\fBftwx\fR visits a directory before visiting any of the files in
the directory. The tree walk continues until the tree is exhausted,
the user function returns a negative value or some error occurs
(since -1 indicates an error in \fBftwx()\fR, the user function
should not return that value).
If the user function returns a positive value for a directory, that
directory is not traversed.
Symbolic links are followed if the \fIflag\fR is \fBFTWX_FOLLOW\fR.
.LP
The \fBdepth\fR argument determines how deep in the tree to go. The
original \fIpath\fR is at depth 0. If \fIdepth\fR is \fBFTWX_ALL\fR,
there is no depth limit. However since \fBftwx()\fR uses a file
descriptor for each level of the tree, there is a limit to
the tree depth that it can process that depends on the number
of available file descriptors.
.SH "RETURN VALUES"
.LP
\fBftwx()\fR
returns 0 if is successful, -1 if an error occurs (\fIerrno\fR is set)
and the (negative) value returned by the user function if that occurs.
.SH "SEE ALSO"
.LP
ftw(3)


ChangeLog
2003-07-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (_fsm_free): Prevent infinite loops on fsm_free
	of corrupted memory

2003-07-09  James Tanis  <jtt@hawking.sysdetect.com>

	* src/dict/dll.c (dll_error_reason): Export pulibcally

2003-06-13  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h, include/ht.h, include/pq.h: avoid stupid void * #1311

2003-06-07  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_create): argsp may change size between
	external and internal users due to HASH_STATS define.

2003-05-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_destroy): Allow clean up of all memory when
	insure is in use and fsma is replaced by malloc

2003-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_STATS): keep performance stats for hash tables

	* src/dict/htimpl.h (struct ht_stats): hash table performance stats
	(ENTRY_IS_FULL): kludge to detect bucket collisions

	* src/dict/ht.c: instrumented to keep performance stats for hash tables
	(print_hashstats): print performance stats

2003-04-28  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/ht.c (ht_delete): microscopic performance tweak
	[Seth got the big win already]

2003-04-16  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Only allow users to
	actually alter the value of thread_overried if we are compiled
	with threads.

2003-04-08  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Provide interface for
	people to specify this will be a threaded program and to enforce
	thread safe memory protection.

2003-04-03  Seth Robertson  <seth@startide.sysdetect.com>

	* src/pq/hpq.c: Make thread-safe, through PQ_THREADED_SAFE

2003-03-31  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/*.c src/fsma/*.c: Make thread-safe, through
	DICT_THREADED_SAFE and DICT_THREADED_MEMORY

2002-12-10  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/pq/Makefile: move -lpq to -lpqueue so as not to conflict
	with postgres

2002-11-11  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/fsma/fsma.c, include/dict.h: opx_rollup_3 merge

2002-11-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* include/dict.h, src/fsma/fsma.c: trheads merge

2002-11-06  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/fsma/fsma.c: Compile problems when using dmalloc

2002-11-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* include/dict.h: Define symbol for NOCOALESCE which makes it
	clear the only reason we are using it is because we are too lazy
	to make CLC thread-safe

2002-09-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/dict/ht.c: insight nits

2002-09-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/dict/dll.c, ChangeLog: Do not create dlls w/o UNORDERED and
	w/o OOCMP function--coredump otherwise sometimes

2002-09-05  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/dll.c (dll_create): Default order is ordered.

	* src/dict/dll.c (dll_create): Use bitwise and for bits, instead
	of addition.  Prevented an argument sanity check from occurring.

2002-08-29  Seth Robertson  <seth@startide.sysdetect.com>

	* include/fsma.h (struct __fsma_header): AHHHHH-HHHHAAAAAAAA!
	The reference counter was previously a short.  Worker, for reasons
	too complicated to get into, can have more than 2^16th lists.
	If we happen (not unlikely) to destroy down to 2^16 (e.g. zero)
	fsma thinks all allocators are free and promptly proceeds to return
	the memory to the system.  Memory was free'd but still in use.
	BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA.

	* src/fsma/fsma.c (fsm_destroy): Perform assertion to verify we
	actually found the fsma allocator in the coalesce list.  Don't
	bother to realloc the coalesce list size down--it won't do no good
	nohow.

2002-08-27  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Serious and quite possibly
	bug-injecting misuse of a hint (cur_min).  The insert and delete
	routines were setting this to a bucket chain entry, not a table
	entry bucket (except when the insert routine had to perform a
	bucket chain expansion--when cur_min might be set to some other
	random value).  This usually was OK since bucket chain numbers are
	very low (0-3 range), but if a valid entry was in the table at
	slot zero or one, then ht_minimum might never see that item.

2002-08-14  James Tanis  <jtt@saint.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Use cached minimum index as search start.
	(ht_do_insert): Update minimum index if less the current minimum.
	(ht_delete): Search for new minimum index if the index of the
	deleted node equals the current minimum.
	In addition add and obj_cnt so we don't try to perform expensive
	operations on empty hashes.

2002-05-13  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): Shared libfsma no longer
	breaks with libtool (on both linux and bsd).

2002-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): back out shared libfsma;
	this breaks libtool in a strange way that requires investigation

2002-05-10  James Tanis  <jtt@winwood.sysdetect.com>

	* src/fsma/fsma.c: Formatting and simplify coalesce code.

	* src/fsma/Makefile (BK_LARGE_LIB): Make libfsma shared.

	* src/dict/bst.c (bst_redelete): Improve efficiency by avoiding
	unnecessary checks and recusive calls.

2002-02-22  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_NUKE): delete contents and dict itself
	(_DICT_NUKE_WORK_LOOP): common code for DICT_NUKE, DICT_NUKE_CONTENTS

	* src/dict/dictimpl.h (HANDLE_ERROR): Eliminate unused id argument
	[use BK_FUNCNAME  here instead if you need it for debugging]

2002-01-31  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/Makefile (BK_LARGE_LIB): use ${LIBEXT} for libtool build

2002-01-09  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/fsma.c: INSIGHT->INSURE


src/misc/ftwx.c
/*
 * (c) Copyright 1992 by Panagiotis Tsirigotis
 * All rights reserved.  The file named COPYRIGHT specifies the terms
 * and conditions for redistribution.
 */


#include <sys/types.h>
#include <sys/stat.h>
#ifndef _WIN32
#include <sys/file.h>
#endif /* !_WIN32 */
#include <stdlib.h>
#include <unistd.h>

#if HAVE_DIRENT_H
# include <dirent.h>
# define NAMLEN(dirent) strlen((dirent)->d_name)
#else
# define dirent direct
# define NAMLEN(dirent) (dirent)->d_namlen
# if HAVE_SYS_NDIR_H
#  include <sys/ndir.h>
# endif
# if HAVE_SYS_DIR_H
#  include <sys/dir.h>
# endif
# if HAVE_NDIR_H
#  include <ndir.h>
# endif
#endif

#if defined (_WIN32) && !defined(__CYGWIN__)
# include <sosnt.h>
#endif /* _WIN32 && !__CYGWIN__ */

extern int errno ;

#include "misc.h"
#include "ftwx.h"
#include "clchack.h"

#define PRIVATE			static

#define NUL					'\0'


typedef enum { NO, YES } boolean_e ;

static struct
{
	int (*stat_func)() ;
	int (*user_func)() ;
} ftwx_data ;

PRIVATE int ftwx_traverse(char *, int depth);

/*
 * ftwx is an extension to ftw, that optionally follows symlinks (the
 * default is NOT to follow them).
 *
 * Possible flag values:
 *		FTWX_FOLLOW: 		follow symlinks
 *
 * Possible depth values:
 *		0			: means only the specified path
 *		positive : means go as deep as specified
 *		FTWX_ALL : no depth limitation
 *
 * User function return value:
 *		negative : means an error occured and the traversal should stop
 *		0			: OK
 *		positive : means that if the current object is a directory it
 *					  should not be traversed.
 *
 * Return value:
 *		-1 		: if an error occurs
 *  frv			: frv is the function return value if it is negative (it
 *					  should not be -1).
 *		0			: if successful
 */
int ftwx(char *path, int (*func) (/* ??? */), int depth, int flags)
{
#ifndef _WIN32
	int stat(const char *, struct stat *), lstat(const char *, struct stat *) ;
#endif /* _WIN32 */

	/*
	 * Initialize the data structure
	 */
	ftwx_data.stat_func = ( flags & FTWX_FOLLOW ) ? stat : lstat ;
	ftwx_data.user_func = func ;

	return( ftwx_traverse( path, depth ) ) ;
}




/*
 * ftwx_traverse works in two phases:
 *
 * Phase 1: process the current path
 *
 * Phase 2: if the current path is a directory, it invokes ftwx_traverse
 *				for each directory entry
 */
PRIVATE int ftwx_traverse(char *path, int depth)
{
	DIR *dirp ;
	struct stat st ;
	int ftw_flag = 0 ;
	boolean_e traverse = YES ;
	int retval ;
	int save_errno ;

	if ( (*ftwx_data.stat_func)( path, &st ) == -1 )
		ftw_flag = FTW_NS ;
	else
	{
		/*
		 * If it is a directory and determine if it is readable
		 * (if it is not readable, we don't traverse it
		 */
		if ( S_ISDIR( st.st_mode ) )
			if ( access( path, R_OK ) == 0 )
				ftw_flag = FTW_D ;
			else
				ftw_flag = FTW_DNR ;
		else
			ftw_flag = FTW_F ;
	}
	retval = (*ftwx_data.user_func)( path, &st, ftw_flag ) ;
	if ( retval < 0 )
		return( retval ) ;
	else if ( retval > 0 && ftw_flag == FTW_D )
		traverse = NO ;

	/*
	 * Stop traversal if:
	 *		a. depth reached 0
	 *		b. the current path is not a readable directory
	 *		c. the user doesn't want us to traverse this directory tree
	 */
	if ( depth == 0 || ftw_flag != FTW_D || traverse == NO )
		return( 0 ) ;

	if ( depth != FTWX_ALL )
		depth-- ;

	if ( ( dirp = opendir( path ) ) == NULL )
		return( -1 ) ;

	for ( ;; )
	{
		struct dirent *dp ;
		char *filename ;

		errno = 0 ;			/* to detect readdir errors */
		dp = readdir( dirp ) ;
		if ( dp == NULL )
		{
			retval = ( errno == 0 ) ? 0 : -1 ;
			break ;
		}

		/*
		 * The special names: "." and ".." are skipped
		 */
		if ( dp->d_name[ 0 ] == '.' )
			if ( dp->d_name[ 1 ] == NUL ||
			     (dp->d_name[ 1 ] == '.' && dp->d_name[ 2 ] == NUL) )
				continue ;

		filename = make_pathname( 2, path, dp->d_name ) ;
		if ( filename == NULL )
		{
			retval = -1 ;
			break ;
		}

		retval = ftwx_traverse( filename, depth ) ;
		free( filename ) ;

		/*
		 * Check for a negative value instead of -1 because the
		 * user function may use any negative value
		 */
		if ( retval < 0 )
			break ;
	}
	/*
	 * Make sure we don't trash errno; we should only do this if
	 * retval is negative, but we are lazy...
	 */
	save_errno = errno ;
	(void) closedir( dirp ) ;
	errno = save_errno ;
	return( retval ) ;
}

